File: app/src/main/java/com/azizbekian/spyur/injection/AppModule.java
package com.azizbekian.spyur.injection;

import android.content.Context;

import com.azizbekian.spyur.api.ApiInteractor;
import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestManager;

import javax.inject.Singleton;

import com.azizbekian.spyur.BuildConfig;
import com.azizbekian.spyur.manager.SpyurManager;
import com.azizbekian.spyur.converter.ListingConverter;
import com.azizbekian.spyur.converter.SearchConverter;
import com.azizbekian.spyur.rest.SpyurApi;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;

/**
 * Created on May 09, 2016.
 *
 * @author Andranik Azizbekian (andranik.azizbekyan@gmail.com)
 */
@Module
public class AppModule {

    private Context context;

    public AppModule(Context context) {
        this.context = context;
    }

    @Provides @Singleton
    public Context provideContext() {
        return context;
    }

    @Provides @Singleton
    public OkHttpClient provideOkHttpClient() {
        OkHttpClient.Builder builder = new OkHttpClient.Builder();

        if (BuildConfig.DEBUG) {
            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
            builder.addInterceptor(interceptor).addInterceptor(chain -> {
                // Simulating network delay
                // SystemClock.sleep(3000);
                return chain.proceed(chain.request());
            });
        }

        return builder.build();
    }

    @Provides @Singleton
    public SpyurApi.SearchApi provideSearchApi(OkHttpClient client) {

        return new Retrofit.Builder()
                .baseUrl(SpyurApi.ENDPOINT)
                .client(client)
                .addConverterFactory(new SearchConverter.Factory())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .build()
                .create(SpyurApi.SearchApi.class);
    }

    @Provides @Singleton
    public SpyurApi.ListingApi provideListingApi(OkHttpClient client) {

        return new Retrofit.Builder()
                .baseUrl(SpyurApi.ENDPOINT)
                .client(client)
                .addConverterFactory(new ListingConverter.Factory())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .build()
                .create(SpyurApi.ListingApi.class);
    }

    @Provides @Singleton
    public ApiInteractor provideSpyurManager(SpyurApi.SearchApi searchApi, SpyurApi.ListingApi
            listingApi) {

        return new SpyurManager(searchApi, listingApi);
    }

    @Provides @Singleton
    public RequestManager provideGlide(Context context) {
        return Glide.with(context);
    }

}

File: app/src/main/kotlin/com/incentive/yellowpages/injection/module/AppModule.kt
package com.incentive.yellowpages.injection.module

import android.content.Context
import com.incentive.yellowpages.data.remote.ApiContract
import com.incentive.yellowpages.injection.ApplicationContext
import com.incentive.yellowpages.misc.converter.ListingConverter
import com.incentive.yellowpages.misc.converter.SearchConverter
import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import java.util.concurrent.TimeUnit
import javax.inject.Singleton

@Module
class AppModule(private val context: Context) {

    companion object {
        val DURATION_CONNECTION_TIMEOUT = 10
        val DURATION_WRITE_TIMEOUT = 10
        val DURATION_READ_TIMEOUT = 30
    }

    @Provides
    @Singleton
    @ApplicationContext
    fun provideAppContext(): Context {
        return context
    }

    @Provides
    @Singleton
    fun applyTimeouts(okHttpBuilder: OkHttpClient.Builder): OkHttpClient {
        return okHttpBuilder.connectTimeout(DURATION_CONNECTION_TIMEOUT.toLong(), TimeUnit.SECONDS)
                .writeTimeout(DURATION_WRITE_TIMEOUT.toLong(), TimeUnit.SECONDS)
                .readTimeout(DURATION_READ_TIMEOUT.toLong(), TimeUnit.SECONDS)
                .build()
    }

    @Provides
    @Singleton
    fun providesSearchApi(okHttp: OkHttpClient): ApiContract.SearchApi {
        return Retrofit.Builder()
                .baseUrl(ApiContract.BASE)
                .client(okHttp)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(SearchConverter.Factory())
                .build().create(ApiContract.SearchApi::class.java)
    }

    @Provides
    @Singleton
    fun providesListingApi(okHttp: OkHttpClient): ApiContract.ListingApi {
        return Retrofit.Builder()
                .baseUrl(ApiContract.BASE)
                .client(okHttp)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(ListingConverter.Factory())
                .build().create(ApiContract.ListingApi::class.java)
    }

}
